sentry_streams/sentry_streams/pipeline/pipeline.py

```diff
@@ -198,3 +216,53 @@ class Reduce(WithInput, Generic[MeasurementUnit, InputType, OutputType]):
     aggregate_backend: Optional[AggregationBackend[OutputType]] = None
     group_by_key: Optional[GroupBy] = None
     step_type: StepType = StepType.REDUCE
+
+
+@dataclass
+class Batch(Reduce, Generic[MeasurementUnit, InputType]):
+    """
+    A step to Batch up the results of the prior step.
+
+    Batch can be configured via batch size, which can be
+    an event time duration or a count of events.
+    """
+
+    # TODO: Use concept of custom triggers to close window
+    # by either size or time
+    batch_size: MeasurementUnit
+    step_type: StepType = StepType.REDUCE
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        self.windowing: TumblingWindow[MeasurementUnit] = TumblingWindow(self.batch_size)
+        self.aggregate_fn = BatchBuilder[InputType]
+
+
+@dataclass
+class FlatMapStep(WithInput):
+    """
+    A generic step for mapping and flattening (and therefore alerting the shape of) inputs to
+    get outputs. Takes a single input to 0...N outputs.
+    """
+
+    step_type: StepType = StepType.FLAT_MAP
+
+
+@dataclass
+class FlatMap(FlatMapStep, TransformStep[Any]):
+    """
+    A FlatMap with a user-defined function.
+    """
+
+
+@dataclass
+class Unbatch(FlatMapStep, TransformFunction[T]):
+    """
+    A step to flatten a batch representation to output its individual elements.
+    """
+
+    @property
+    def resolved_function(
+        self,
+    ) -> Callable[..., T]:
+        return cast(Callable[..., T], unbatch)
```

Comment:
I wanted to link the InputType of a Batch to the output type of an Unbatch, so that they are forced to be the same (assuming this is the semantics we want to support). 

Thoughts?